File: persona1_java/src/cc4p1/raft/LogEntry.java
package cc4p1.raft;

import cc4p1.common.ProtocolMessage;


public class LogEntry {

    public final long index;
    public final long term;
    public final int camera;
    public final String clase;
    public final String timestamp;
    public final String imagePath;

    public LogEntry(long index, long term, int camera, String clase, String timestamp, String imagePath) {
        this.index = index;
        this.term = term;
        this.camera = camera;
        this.clase = clase;
        this.timestamp = timestamp;
        this.imagePath = imagePath;
    }

    public String encode() {
        String s = ProtocolMessage.ENTRY_FIELD_SEP;
        return index + s + term + s + camera + s + clase + s + timestamp + s + imagePath;
    }

    public static LogEntry decode(String raw) {
        String[] f = raw.split(ProtocolMessage.ENTRY_FIELD_SEP, 6);
        return new LogEntry(
                Long.parseLong(f[0]),
                Long.parseLong(f[1]),
                Integer.parseInt(f[2]),
                f[3],
                f[4],
                f[5]
        );
    }

    @Override
    public String toString() {
        return "#" + index + " (term " + term + ") camara=" + camera + " clase=" + clase
                + " hora=" + timestamp + " img=" + imagePath;
    }
}

File: persona2_python/src/raft/log_entry.py
"""
Una entrada del log replicado: un "encuentro" (evento de deteccion de
objeto/animal/persona). El formato de codificacion DEBE coincidir con
cc4p1.raft.LogEntry (Java):

    index,term,camera,clase,timestamp,imagePath

con las entradas completas separadas entre si por "~" (ProtocolMessage.ENTRY_SEP).
"""

from dataclasses import dataclass

from common import protocol


@dataclass
class LogEntry:
    index: int
    term: int
    camera: int
    clase: str
    timestamp: str
    image_path: str

    def encode(self) -> str:
        s = protocol.ENTRY_FIELD_SEP
        return s.join([
            str(self.index), str(self.term), str(self.camera),
            self.clase, self.timestamp, self.image_path,
        ])

    @staticmethod
    def decode(raw: str) -> "LogEntry":
        # limit=5 -> el 6to campo (image_path) puede contener comas si algun
        # dia se usan rutas con comas; se deja el resto como un solo campo.
        partes = raw.split(protocol.ENTRY_FIELD_SEP, 5)
        return LogEntry(
            index=int(partes[0]),
            term=int(partes[1]),
            camera=int(partes[2]),
            clase=partes[3],
            timestamp=partes[4],
            image_path=partes[5],
        )

    def __str__(self):
        return "#%d (term %d) camara=%d clase=%s hora=%s img=%s" % (
            self.index, self.term, self.camera, self.clase, self.timestamp, self.image_path
        )
